### PR TITLE
Add Typedoc Document Generator

### PR DESCRIPTION
**Pull request by jLouzado**
# Changes
- Documentation now available at https://action-land.github.io/action-land/
- Build local documentation with `yarn build:docs`

### CODE REVIEW

**Comment by tusharmath** [comment]
Docs need not be committed to git. Use travis gh pages deploy functionality to publish typedoc output.

File: .gitignore [hunk 1]
```diff
@@ -6,4 +6,5 @@ node_modules/
 *.map
 .idea/
 *.log
-.vscode/
```
[the remark this answers is not in the document]
Comment (jLouzado) [reply .gitignore]: - It's already been added in `master` actually, it's just repeated here.
- Created a separate PR to organize `.gitignore` also, if that's more helpful
  - anyway `vscode` is the only additional IDE

File: tsconfig.json [hunk 2]
```diff
@@ -3,5 +3,15 @@
...
+  },
+  "typedocOptions": {
+    "mode": "file",
+    "out": "docs",
+    "externalPattern": "**/node_modules",
+    "excludeExternals": true,
+    "excludeNotExported": true,
+    "name": "@action-land",
+    "readme": "./README.md",
+    "toc": true
```
[the remark this answers is not in the document]
Comment (jLouzado) [reply tsconfig.json]: - From the documentation it's required for rendering the `Table Of Contents` on the right:

image

- However, setting it to `false` doesn't seem to produce any visible change. I've raised [an issue]( https://github.com/TypeStrong/typedoc/issues/1040) with the typedoc team, and for now I'll remove the option.

File: .travis.yml [hunk 3]
```diff
@@ -14,7 +14,19 @@ script:
   - yarn lint
   - yarn tsc
   - yarn test
+  if: branch = master
+  - yarn build:docs
 before_install: yarn global add greenkeeper-lockfile@2
 before_script: greenkeeper-lockfile-update
 after_script: greenkeeper-lockfile-upload
 after_success: ./deploy.sh
+deploy:
```
[the remark this answers is not in the document]
Comment (jLouzado) [reply .travis.yml]: - `deploy` is not a hook as such, it's a "phase" in the Job Lifecycle
- `script` denotes a build phase, after that there's an optional phase where you can "deploy" some build product to one of the available providers (Heroku, AWS or GithubPages, in our case)
- `deploy` is skipped if there's an error during the "script" phase

ref: https://docs.travis-ci.com/user/job-lifecycle/#deploying-your-code

File: .gitignore (hunk 1 above)
[the remark this answers is not in the document]
Comment (jLouzado) [reply .gitignore]: This is just a newline added, probably by prettier.

[the remark this answers is not in the document]
Comment (jLouzado) [reply .gitignore]: .vscode is already added in `master`, this was just adding the `doc` folder to the gitignore since we don't want it to be committed

**Comment by Gerrit0** [comment]
Question - where did you see that `toc` was a Boolean option? It is an array, which by default is filled with all root symbols. If you set it to a smaller array, it will limit the toc to those elements.

File: .travis.yml (hunk 3 above)
Comment (fenilkanjani) [inline .travis.yml]: I think you should use [these](https://docs.travis-ci.com/user/build-stages/#specifying-stage-order-and-conditions) conditions

Comment (jLouzado) [reply .travis.yml]: @fenilkanjani check this out, see if it makes sense. These are the specific instructions for deploying to github pages.

https://docs.travis-ci.com/user/deployment/pages/

The options you shared are for defining a custom deploy stage, here we're passing options to the `github pages` provider itself, which is taking care of the deployment

File: .gitignore (hunk 1 above)
[the remark this answers is not in the document]
Comment (jLouzado) [reply .gitignore]: done in #27

File: .travis.yml (hunk 3 above)
Comment (fenilkanjani) [reply .travis.yml]: What I meant was, should not we run this conditionally with:
```
stages:
  - compile
  - test
  - name: deploy
    if: branch = master
```

Comment (jLouzado) [reply .travis.yml]: The conditional behaviour is already there, it's specified at the line below. Let's discuss this in-person after standup?

https://github.com/action-land/action-land/pull/25/files#diff-354f30a63fb0907d4ad57269548329e3R31

Comment (jLouzado) [reply .travis.yml]: - As long as our build is a multi-job (unlike PWA), then it's fine to just use the `deploy` hook to update documentation.

@fenilkanjani Documentation for the `on` flag in the `deploy` life-cycle method.

> When all conditions specified in the on: section are met, your build will deploy.

https://docs.travis-ci.com/user/deployment/#conditional-releases-with-on

File: .travis.yml [hunk 4]
```diff
@@ -14,7 +14,19 @@ script:
...
+  provider: pages
+  skip_cleanup: true
+  local_dir: './docs'
+  keep_history: true
```
Comment (tusharmath) [inline .travis.yml]: should be `false`. No point keeping history when you can re-create it from source code.

File: .travis.yml [hunk 5]
```diff
@@ -14,7 +14,19 @@ script:
...
+  project_name: '@action-land'
+  github_token: $GH_TOKEN
+  verbose: true
```
Comment (tusharmath) [inline .travis.yml]: `verbose: false`

File: tsconfig.json [hunk 6]
```diff
@@ -3,5 +3,14 @@
...
     "strict": true
+  },
+  "typedocOptions": {
+    "mode": "file",
+    "out": "docs",
+    "externalPattern": "**/node_modules",
+    "excludeExternals": true,
+    "excludeNotExported": true,
+    "name": "@action-land",
+    "readme": "./README.md"
```
Comment (tusharmath) [inline tsconfig.json]: Makes sense to have a separate file `typedoc.json`

```json
{
  "excludeExternals": false,
  "excludeNotExported": true,
  "excludePrivate": true,
  "excludeProtected": true,
  "hideGenerator": true,
  "mode": "file",
  "out": "docs",
  "exclude": ["test/**"],
  "theme": "default",
  "listInvalidSymbolLinks": true
}
```

**Pull request merged**